pkg/algorithm/utils.go

```diff
@@ -0,0 +1,449 @@
+		}
+		klog.Infof("[%v]: scheduled to node %v, GPUs %v",
+			internal.Key(pod), selectedNode, common.ToJson(selectedGpuIndices))
+		return internal.PodScheduleResult{
+			PodBindInfo: &api.PodBindInfo{
+				Node:                  selectedNode,
+				GpuIsolation:          selectedGpuIndices,
+				CellChain:             cellChain,
+				AffinityGroupBindInfo: affinityGroupBindInfo,
+			},
+		}
+	}
+}
+
+// generateAffinityGroupBindInfo translates the physical and virtual placements of an affinity group
+// into a a series of AffinityGroupMemberBindInfos, and also returns the allocated node and GPU addresses
+// of the current pod.
+func generateAffinityGroupBindInfo(
+	groupPhysicalPlacement groupPhysicalPlacement,
+	groupVirtualPlacement groupVirtualPlacement,
+	cellLevelToType map[CellChain]map[CellLevel]api.CellType,
+	currentGpuNum int32,
+	currentPodIndex int32,
+	group *AlgoAffinityGroup,
+	groupName string,
+	suggestedNodes common.Set) (
+	affinityGroupBindInfo []api.AffinityGroupMemberBindInfo,
+	nodesNotInSuggested []string,
+	selectedNode string,
+	selectedGpuIndices []int32,
+	chain string) {
+
+	if groupPhysicalPlacement == nil {
+		return
+	}
+	affinityGroupBindInfo = make([]api.AffinityGroupMemberBindInfo, len(groupPhysicalPlacement))
+	groupMemberIndex := 0
+	for podGpuNum, podPhysicalPlacements := range groupPhysicalPlacement {
+		mbi := api.AffinityGroupMemberBindInfo{
+			PodPlacements: make([]api.PodPlacementInfo, len(podPhysicalPlacements)),
+		}
+		for podIndex := int32(0); podIndex < int32(len(podPhysicalPlacements)); podIndex++ {
+			mbi.PodPlacements[podIndex].PhysicalGpuIndices = make([]int32, podGpuNum)
+			mbi.PodPlacements[podIndex].PreassignedCellTypes = make([]api.CellType, podGpuNum)
+			for gpuIndex := int32(0); gpuIndex < podGpuNum; gpuIndex++ {
+				pGpu := podPhysicalPlacements[podIndex][gpuIndex]
+				if pGpu == nil {
+					if group == nil {
+						panic(fmt.Sprintf("The first pod in group %v was allocated invalid resource", groupName))
+					}
+					// if the physical placement of this pod is not found (e.g., removed due to reconfiguration),
+					// we will insist the decision by retrieving it from other pods
+					mbi.PodPlacements[podIndex], chain = retrieveMissingPodPlacement(group, podGpuNum, podIndex)
+					klog.Warningf(
+						"pod placement has been invalid and is retrieved from annotation of other pods: node %v, GPU %v",
+						mbi.PodPlacements[podIndex].PhysicalNode, mbi.PodPlacements[podIndex].PhysicalGpuIndices[gpuIndex])
+				} else {
+					nodes, gpuIndices := pGpu.(*PhysicalCell).GetPhysicalPlacement()
+					// here each cell (i.e., pGpu) is only one GPU, hence we takes the first element
+					// in its "nodes" and "gpuIndices" as the node and GPU address
+					if mbi.PodPlacements[podIndex].PhysicalNode == "" {
+						mbi.PodPlacements[podIndex].PhysicalNode = nodes[0]
+						if !suggestedNodes.Contains(nodes[0]) {
+							nodesNotInSuggested = append(nodesNotInSuggested, nodes[0])
```

Comment:
>nodesNotInSuggested [](start = 7, length = 19)

Use set instead of slice, due to in log we will see:

not within K8s suggested nodes: ["10.151.41.23","10.151.41.23","10.151.41.23"]